p2p/p2p.go

```diff
@@ -1,61 +1,186 @@
+			if err := p2p.store.Write(multi.Addr().String(), multi); err != nil {
+				p2p.logger.Errorf("cannot add multi-address to store: %v", err)
+				return
+			}
+			if response.Error != nil {
+				p2p.logger.Warnf("received error in response: code = %v, message = %v, data = %v", response.Error.Code, response.Error.Message, string(response.Error.Data))
+				return
+			}
+
+			// Parse the response and write any multi-addresses returned by the node to the store.
+			var result jsonrpc.QueryPeersResponse
+			if err := json.Unmarshal(response.Result, &result); err != nil {
+				p2p.logger.Errorf("invalid QueryPeersResponse from node %v: %v", multi.Addr().String(), err)
+				return
+			}
+			for _, node := range result.Peers {
+				multiAddr, err := peer.NewMultiAddr(node, 0, [65]byte{})
+				if err != nil {
+					p2p.logger.Errorf("invalid QueryPeersResponse from node %v: %v", multi.Addr().String(), err)
+					return
+				}
+				if err := p2p.store.Write(multiAddr.Addr().String(), multiAddr); err != nil {
+					p2p.logger.Errorf("cannot add multi-address to store: %v", err)
+					return
+				}
+			}
+		})
+		p2p.logger.Infof("connecting to %v darknodes", p2p.store.Entries())
+	}()
+
+	return nil
 }
 
-type QueryAddressRequest struct {
-	// TODO: Use decorator pattern.
-	Addr addr.Addr
+// handleQuery retrieves multi-addresses from the store and writes the response to the responder channel in the original
+// request.
+func (p2p *P2P) handleQuery(message rpc.QueryMessage) tau.Message {
+	var response jsonrpc.Response
+	var responder chan<- jsonrpc.Response
+	switch request := message.Request.(type) {
+	case jsonrpc.QueryPeersRequest:
+		// Return at most 5 random addresses from the multi-address store.
+		iter := p2p.store.Iterator()
+		addresses := p2p.randomPeers(iter, 5)
+		response = jsonrpc.QueryPeersResponse{
+			Peers: addresses,
+		}
+		responder = request.Responder
+	case jsonrpc.QueryNumPeersRequest:
+		// Return the number of entries in the store.
+		entries := p2p.store.Entries()
+		response = jsonrpc.QueryNumPeersResponse{
+			NumPeers: entries,
+		}
+		responder = request.Responder
+	default:
+		p2p.logger.Panicf("unknown query request type: %T", request)
+	}
+
+	if responder == nil {
+		// This is a defensive check.
+		p2p.logger.Error("query responder channel is nil")
+	} else {
+		responder <- response
+	}
+
+	return nil
 }
 
-func (message QueryAddressRequest) IsMessage() {
+// randomPeers returns at most 5 random multi-addresses from the store.
+func (p2p *P2P) randomPeers(iter store.KVStoreIterator, num int) []string {
+	// Retrieve all the Darknode multi-addresses in the store.
+	addresses := make([]string, 0, p2p.store.Entries())
+	for iter.Next() {
+		var value peer.MultiAddr
+		_, err := iter.KV(&value)
+		if err != nil {
+			p2p.logger.Errorf("cannot read multi-address using iterator: %v", err)
+			continue
+		}
+		addresses = append(addresses, value.Value())
+	}
+
+	// Shuffle the list.
+	rand.Seed(time.Now().UnixNano())
```

Comment:
You only need to seed the randomiser once at the beginning of the program. This is usually done at the beginning of the main function (or test). It is better to leave seeding the randomise to the user because they might want reproducibility between runs (and therefore may not want to seed).